### ISSUE TITLE
get rid of most of `storeDir` - the troublemaker

### ISSUE THREAD

**Issue by sergpolly** [mirnylab/distiller-nf#52]
`storeDir` is a huge troublemaker in a cluster environment: whenever some process fails downstream, for whatever reason,  some upstream processes can be copying intermediate results to `storeDir` location. After the restart or `-resume` processes take off with whatever is present in the `storeDir` - which isn't always a good thing, as results might be incomplete or whatever - it often creates a mess.

 - Jobs/processes could fail on a cluster for many reasons which we could not control and even using `errorStrategy='retry'` does not necessarily saves us in all cases.
 - `storeDir` intended use is to store things as a permanent cache, in most cases after downloading from some database (like SRA), but not to store all intermediate results there (according to Paolo)

**Comment by sergpolly** [comment]
from the times that I had trouble with StoreDir: https://github.com/nextflow-io/nextflow/issues/406

**Comment by sergpolly** [comment]
`storeDir` behavior have been fixed apparently:
https://github.com/nextflow-io/nextflow/issues/227

also review this to see if it's relevant: https://github.com/nextflow-io/nextflow/issues/592

**Comment by sergpolly** [comment]
just would llike to document some of the evidence of bad `storeDir` behavior:
from the recent distiller run, using `sra_download` process, which is using `storeDir` (the recommended way, actually):
![screenshot from 2018-12-06 18-32-46](https://user-images.githubusercontent.com/6790270/49618227-c5d77b00-f985-11e8-8c91-216582bcf462.png)
so , seems like 1 `sra_download` failed and re-submitted - let's see, what was inside of the corresponding `work` folders:
![screenshot from 2018-12-06 18-37-51](https://user-images.githubusercontent.com/6790270/49618303-14851500-f986-11e8-8377-db09ee49714d.png)

actually both folders looks identical and just fine!
apparently something happened during the transfer ...

**Comment by golobor** [comment]
I don't think I follow what's the issue here. Could you please elaborate?

**Comment by sergpolly** [comment]
my last post is not a new issue, is just an illustration for an ongoing "discussion" about `storeDits` vs `publishDirs`

it just illustrates, how exactly `storeDirs` fails to copy stuff over to `intermediates`, while the corresponding process itself finished fine and all of the files are intact inside of the corresponding `work` folder

I still don't understand why is it the transfer from `work` to `intermediates` that  fails quite often for us on the cluster

**Comment by sergpolly** [comment]
On the same topic , form the recent conversation with @mimakaev : 

> Also on the topic of this nextflow "bug" - what I meant by that is that copying to `intermediates` looks strange to me (screenshot attached) - why do they add this `||true` thing at the end of the cp command ?! it seems to me like the success of copying to `intermedaites` does not affect anything ... - i'm not a bash expert don't know for sure.
You can check it in any of your distiller runs inside of `work/hash-id/.command.run` script at the bottom.


![screenshot from 2018-12-07 18-39-18](https://user-images.githubusercontent.com/6790270/49679346-99406380-fa58-11e8-89df-c8c9b8efce61.png)

**Comment by golobor** [comment]
issue https://github.com/mirnylab/distiller-nf/issues/57 has been merged into this one

**Comment by sergpolly** [comment]
Yes, yes - nothing major , just a wasteful re-execution.
It almost seems contradictory, as the process finished, ok - files were generated in work, but then something triggered re-execution. Was it faulty transfers or something else?- don't know. It shouldn't have due to ||true at the end of the copy command ?!
I'm still struggling to understand the "||true" bash syntax

Regrardless, ErrorStrategy=retry is a huge helper in our execution environment

**Comment by golobor** [comment]
so, the || true bash syntax guarantees that the process does not return an
error (error code 0) even if the main process actually failed.
E.g. rm non-existent-file  would give you an error, but ( rm
non-existent-file ) || true gives you 0, the sae way false || true == true.

**Comment by sergpolly** [comment]
so, finally ! ladies and gentelmen - a beautiful evidence that `nextflow` does not give an f about transfer-success to `intermediates` (just as @golobor explained about `||true`)  - we knew about all of that mostly - but it would be nice to document it:
![screenshot from 2018-12-13 19-29-23](https://user-images.githubusercontent.com/6790270/49975933-05661000-ff0e-11e8-8e20-1ec84acf9038.png)

above: in a `work` folder we've got 2 beautiful `fastq-s`, left and right, for NPC library from Bonev2017 dataset, about ~8GB each
below: in `intermediates` corresponding files have sizes 2.5 and ~8GB respectively ...

corruption evidence or just faulty transfer evidence is here to stay.
to be fair it's `nextflow version 0.31.1.4886` , not the latest `18 `-something (yes they went from `0.` to `18.` - because of the year or something like that)
@Phlya did you experience corruptions like that before/ever ?

**Comment by Phlya** [comment]
I never compared output with work/, but could easily be that the problem I
had was due to this!

**Comment by mimakaev** [comment]
Is there a nextflow ticket about that?

**Comment by golobor** [comment]
Don't remember seeing one. But we would need to have a reproducible example
to file a ticket :[

Anton.
(typed on a phone, sorry for brevity and typos)

**Comment by mimakaev** [comment]
We don't need one to ask why the return status of file transfer is not being checked/interpreted. 

However, the problem is still very cryptic for me. I don't understand what is leading to it: an error in cp/mv? or premature return from the process before cp/mv finishes? 

@sergpolly  Do we have any way to see the command that was submitted to the cluster?  Is it docker, or a bash script? Do we explicitly see how copying was done? 

Can we maybe replace cp/mv that is being "responsible" for the error the node with an alias that does some logging and pre/post md5 check? 

@golobor  Who does the copying/transfer happen? Something inside docker?

**Comment by sergpolly** [comment]
@mimakaev 
> However, the problem is still very cryptic for me. I don't understand what is leading to it: an error in cp/mv? or premature return from the process before cp/mv finishes?

I'm not 100% sure about this one ... This "symptom" is present only (as far as I my experience goes) when `storeDir` directive is used to store intermediate results. Results that are "published" using `publishDir` are not used used for downstream processing, thus do not trigger any downstream errors, and I've never encountered "published" results that were corrupted or half-copied or anything like that. to be fair, such results are typically smaller - `.pairs`, `.cool`; as opposed to fastq-s stored by `donwload_sra`.

We cannot even agree with @Phlya that we are having the same exact issue ...
Maybe it's just our cluster NFS-nearlines ...
It has to be a "big" thing for the problem to occur - 2009-erez test went fine, without corruption or half-way copy; bonev-2017 - I've seen it 3 times - only for `download_sra`, where I still use `storeDir` (the only place for me - I've been testing with the "old" distiller still, I'm planning to launch "merge_parse_map"-one, once the old is one is done) 

Now to the difference between `storeDri` and `publishDir` and to 
> @sergpolly Do we have any way to see the command that was submitted to the cluster? Is it docker, or a bash script? Do we explicitly see how copying was done?

each `process` of `nextflow` script/pipeline is being executed in an "isolated" environment - i.e. a directory inside of `./work`, `work/short-hash/longer-hash`. Each one of these directories correspond to 1 instance of a process, i.e. mapping of 1 chunk  -> one `work/hash/long-hash` folder . 
Inside of each one of them there are staged input files, output files and a bunch of hidden files (including bash scripts, logs, stdout, exitcode, trace etc.) that control execution and "document" it.
To those bash scripts - there are 3 of them: `.command.run`, `.command.stud` and `.command.sh` - they are nested in this exact order, i.e.
 - `.command.sh` - is the innermost one and it's an actual command that user/pipeline creator wants to be executed - like this bwa-command in triple quotes: https://github.com/mirnylab/distiller-nf/blob/ab5044f6a6bc1478041f4d7f5f355e0321f39279/distiller.nf#L370
 - `.command.stud` one is a little wrapper over `.command.sh`, that from my understanding, simply tracks resource usage
 - finally, `.command.run` is the outermost one - master scripts that wraps `.command.stud`, here things get `executor` dependent, i.e. there would be a difference between `.command.run` between local and cluster execution. Ina case of a cluster `.command.run` contains a typical header with resource requests (long queue, 15 hours, 8 cores on 1 host, 8gb per core, etc ):
![screenshot from 2018-12-16 22-01-59](https://user-images.githubusercontent.com/6790270/50064271-4ebc9680-017e-11e9-8ce2-6aec3e0801e9.png)
i dont' have examples of a local `.command.run` - but you guys should have some for sure - it should be simplier a bit I'd guess.
Then `.command.run` contains some staging commands (lots of bash-magic, as @golobor calls it, I do not fully understand it - just some concepts, not details) , creating symlinks for input data etc, and finally finally finally - it calls that freaking `.command.stud`-thing - that's where we have our docker/singularity or nothing situation (`.command.stud` called inside of local singularity container in the next screenshot):
![screenshot from 2018-12-16 22-10-28](https://user-images.githubusercontent.com/6790270/50064479-73fdd480-017f-11e9-8b27-99c7c52d2499.png)
so `.command.run` runs "natively", outside of docker/singularity , while `.command.stud`, that calls `.command.sh` run inside of the container (if specified).
Now to:
> Do we explicitly see how copying was done?

that depends on the situation - i.e. I can name at least 3 scenarios:
1. if we use `storeDir` - `cp`-ing to intermediates happens in the "native" `.command.run` right after the `.command.stud` call , and only if that call was successful (as far as my knowledge of `bash` magic can tell me), screenshot: one in this comment https://github.com/mirnylab/distiller-nf/issues/52#issuecomment-445413620
https://user-images.githubusercontent.com/6790270/49679346-99406380-fa58-11e8-89df-c8c9b8efce61.png
2. if no `storeDir`/`publishDir` used than , no copying is done - resutls stay right in `work/hash/lognhash` location and are used as input for downstream processing directly (downstream process uses soft-links to the upstream)
3. if `publishDir` is used then ... there is NO copying command visible in either of the 3 `.command./run/stud/sh` - maybe it's done by the master `Java` process - please see orange `Warning` at the end of this paragraph: https://www.nextflow.io/docs/latest/process.html#publishdir
For example see a screenshot of the `work/hash/hash/.command.run` for an instance of `filter_make_pairs` which relies on `publishDirs` to copy stuff to `output/resutls`: 
![screenshot from 2018-12-16 22-35-52](https://user-images.githubusercontent.com/6790270/50065165-05227a80-0183-11e9-8656-4608bb1f3014.png)

**Comment by sergpolly** [comment]
More `nextflow` stuff for reference/whatever:
![screenshot from 2018-12-16 22-46-40](https://user-images.githubusercontent.com/6790270/50065435-89293200-0184-11e9-806d-66d47e18525d.png)
in that example "skipped" processes are those started off of the `intermediates` i.e., stuff that was `storeDir`-ed
"cached" are those resumed from successfuly completed processes in `work/hash/longhash`, here that `[xx/yyyyy]` thing to the left is the `hash/longhash` part of the path to that process execution environment:
`cd work/xx/yyyy`-> [TAB] -> `cd work/xx/yyyyyyyyyyyyyyyyyyyyyyyyy`  should work.



On a positive note, here is an example of why I personally love nextflow:
![screenshot from 2018-12-16 22-53-11](https://user-images.githubusercontent.com/6790270/50065620-795e1d80-0185-11e9-8884-b86fc0a80d22.png)

`140` is an LSF(our cluster's batch system) error code for the time out - i.e. if I asked 3 hours for a task, than it's going to be killed after 3 hours+/- 5 sec of execution - nextflow can resubmit that job dynamically asking for more resources/time ...

**Comment by sergpolly** [comment]
Another piece of evidence to the issue: https://github.com/mirnylab/distiller-nf/issues/52#issuecomment-447173262

That `download_sra` job took almost 4 hours , actually - exactly 4 hours - and it was killed by the scheduler ... but apprently the `.exitcode` was already there , and it was `0` - success - so nextflow proceeded with the incomplete transfer to `intermediates` ...

I thinks it's enough evidence and understanding to open up a `nextflow` issue . 
What do you guys think ?

**Comment by mimakaev** [comment]
I think this is sufficient evidence, and the proposed mechanism is very clear. I think it  should be submitted as a major issue. We can try to replicate it by creating one job that copies a large file.

**Comment by mimakaev** [comment]
I've reproduced it with two simple processes and manual killing

**Comment by mimakaev** [comment]
1. Use example.nf file below
2. Run "cat /dev/urandom > testfile" for a minute or a bit more (so you have enough time to kill copy later)
3. Run nextflow  (nextflow example.nf  --in testfile)
4. Run htop. Observe cp input.fa output_test. It will change to "cp -fRL output_test ...". At that point be quick and kill it 
5. First process finishes successfully, and second process "copyAgain" starts copying a truncated copy.  And output file is a truncated copy of testfile. 

```
#!/usr/bin/env nextflow

params.in = "$HOME/sample.fa"

sequences = file(params.in)

process copyFile {
    storeDir 'teststore'

    input:
    file 'input.fa' from sequences

    output:
    file 'output_*' into records

    """
    cp input.fa output_test
    """

}

process copyAgain {
    
    input:
    file x from records
    
    output:
    file '*.out'

    """
    cat $x  > file.out
    """
}


```
